swig/cpp/src/Session.cpp

```diff
@@ -0,0 +1,281 @@
+{
+    switch(error) {
+    case(SR_ERR_INVAL_ARG):
+        throw runtime_error(sr_strerror(SR_ERR_INVAL_ARG));
+    case(SR_ERR_NOMEM):
+        throw runtime_error(sr_strerror(SR_ERR_NOMEM));
+    case(SR_ERR_NOT_FOUND):
+        throw runtime_error(sr_strerror(SR_ERR_NOT_FOUND));
+    case(SR_ERR_INTERNAL):
+        throw runtime_error(sr_strerror(SR_ERR_INTERNAL));
+    case(SR_ERR_INIT_FAILED):
+        throw runtime_error(sr_strerror(SR_ERR_INIT_FAILED));
+    case(SR_ERR_IO):
+        throw runtime_error(sr_strerror(SR_ERR_IO));
+    case(SR_ERR_DISCONNECT):
+        throw runtime_error(sr_strerror(SR_ERR_DISCONNECT));
+    case(SR_ERR_MALFORMED_MSG):
+        throw runtime_error(sr_strerror(SR_ERR_MALFORMED_MSG));
+    case(SR_ERR_UNSUPPORTED):
+        throw runtime_error(sr_strerror(SR_ERR_UNSUPPORTED));
+    case(SR_ERR_UNKNOWN_MODEL):
+        throw runtime_error(sr_strerror(SR_ERR_UNKNOWN_MODEL));
+    case(SR_ERR_BAD_ELEMENT):
+        throw runtime_error(sr_strerror(SR_ERR_BAD_ELEMENT));
+    case(SR_ERR_VALIDATION_FAILED):
+        throw runtime_error(sr_strerror(SR_ERR_VALIDATION_FAILED));
+    case(SR_ERR_COMMIT_FAILED):
+        throw runtime_error(sr_strerror(SR_ERR_COMMIT_FAILED));
+    case(SR_ERR_DATA_EXISTS):
+        throw runtime_error(sr_strerror(SR_ERR_DATA_EXISTS));
+    case(SR_ERR_DATA_MISSING):
+        throw runtime_error(sr_strerror(SR_ERR_DATA_MISSING));
+    case(SR_ERR_UNAUTHORIZED):
+        throw runtime_error(sr_strerror(SR_ERR_UNAUTHORIZED));
+    case(SR_ERR_LOCKED):
+        throw runtime_error(sr_strerror(SR_ERR_LOCKED));
+    case(SR_ERR_TIME_OUT):
+        throw runtime_error(sr_strerror(SR_ERR_TIME_OUT));
+    }
+}
+
+Session::Session(Connection& sys, const sr_conn_options_t opts, const char *user_name)
+{
+    int rc;
+    _opts = opts;
+    _datastore = SR_DS_STARTUP;
+
+    if (user_name == NULL) {
+        /* start session */
+        rc = sr_session_start(sys.get_conn(), _datastore, _opts, &_sess);
+        if (SR_ERR_OK != rc) {
+            goto cleanup;
+        }
+    } else {
+        /* start session */
+        rc = sr_session_start_user(sys.get_conn(), user_name, _datastore, _opts, &_sess);
+        if (SR_ERR_OK != rc) {
+            goto cleanup;
+        }
+    }
+
+    return;
+
+cleanup:
+    throw_excption(SR_ERR_INIT_FAILED);
+    //TODO error handling
+    return;
+}
+
+int Session::session_stop()
+{
+    int ret = sr_session_stop(_sess);
+    if (ret != SR_ERR_OK) {
+        throw_excption(ret);
+    }
+    return ret;
+}
+
+int Session::get_last_error(Error& err)
+{
+    int ret = sr_get_last_error(_sess, &err.info);
+    if (ret != SR_ERR_OK) {
+        throw_excption(ret);
+    }
+    return ret;
+}
+
+int Session::get_last_errors(Error& err)
+{
+    int ret = sr_get_last_errors(_sess, &err.info, &err.cnt);
+    if (ret != SR_ERR_OK) {
+        throw_excption(ret);
+    }
+    return ret;
+}
+
+int Session::list_schemas(Schema& schema)
```

Comment:
Just an idea: Isn't returning of of the error code unnecessary in methods like this, given that the method will throw an exception if the wrapped function won't succeed? Methods like this could return the output argument instead (the Schema object in this case), which would also simplify the API.